File: group_perm.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void printGroupsRec(char alf[], int r, char *w, int alfIni, int wSize) {
	if (r == 0) {
		printf("%s\n", w);
	} else {
		int n = strlen(alf);
		for(int i=alfIni; i <= (n-r); i++) {
			w[wSize - r] = alf[i];
			printGroupsRec(alf, r-1, w, 1+i, wSize);
		}
	}
}

void printGroups(char alf[], int r) {
	char *w = (char*) malloc(r+1);
	memset(w, 0, r+1);
	printGroupsRec(alf, r, w, 0, r);
	free(w);
}

int main()
{
	printGroups("abcde", 3);
}

File: 2014.2-discretas/group_perm.rb
def print_groups(alf, r, group="")
  n = alf.size
  if r == 0
    puts group
  else
    for i in 0..(n-r)
      new_alf = alf[(i+1)..-1]
      print_groups(new_alf, r-1, group+alf[i])
    end
  end
end

print_groups("abcde", 3)
